### PR TITLE
[FLINK-18579][FLINK-15728][jdbc] Introduce FieldNamedPreparedStatement to support fields are bound multiple times in update statement

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Currently, the update statement from `JdbcDialect` is always broken, because fields will be bound multiple times and this is not supported in the Jdbc source/sinks.

We introudced a `FieldNamedPreparedStatement` support fields are bound multiple times in update statement. Users can use `:name` to reference a value of `name` field, instead of using `?`. So that, users can reference a same field in the statement multiple times easily. For example, an update statement will be:

```sql
UPDATE `tbl` SET `id` = :id, `name` = :name, `email` = :email, `ts` = :ts, 
`field1` = :field1, `field_2` = :field_2
WHERE `id` = :id
```

## Brief change log

- The first commit 164b4a280cbd1083b87cf0cc887efacd8afeea67 remove the deprecated classes, this can reduce the amount of changes.
- Introduce `FieldNamedPreparedStatement`, the implementation is simply mapping the field index to the parameter indexes and forwarding to invoke the underlying `PreparedStatement`.
- Update the new RowData-based source/sink implementations to use the `FieldNamedPreparedStatement`.
- We keep the old source/sink implementation as it is, because this requires a bigger refactoring and the current interface `JdbcStatementBuilder` conflicts with the `FieldNamedPreparedStatement` interface. So old source/sink connectors still have the problem of update statement. 

## Verifying this change

- Added `FieldNamedPreparedStatementImplTest` to test the field name binding.
- Updated `JdbcDynamicTableSinkITCase#testUpsert` to use a smaller batching size which can reproduce the problem. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
cc @leonardBang , appreciate if you can have a look when you have time.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/statement/FieldNamedPreparedStatement.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+ *
+ * <pre>
+ *   Connection con = getConnection();
+ *   String query = "select * from my_table where first_name=:name or last_name=:name";
+ *   FieldNamedPreparedStatement st = FieldNamedPreparedStatement.prepareStatement(con, query, new String[]{"name"});
+ *   st.setString(0, "bob");
+ *   ResultSet rs = st.executeQuery();
+ * </pre>
+ */
+public interface FieldNamedPreparedStatement extends AutoCloseable {
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/statement/FieldNamedPreparedStatement.java:56]: I have considered this. However, the parameters meaning of `setXxx(fieldIndex, x)` is different with `PreparedStatement`. In order to avoiding confusion, I would like to separate them.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java [hunk 2]
```diff
@@ -97,7 +99,7 @@ default String quoteIdentifier(String identifier) {
 	 */
 	default String getRowExistsStatement(String tableName, String[] conditionFields) {
 		String fieldExpressions = Arrays.stream(conditionFields)
-			.map(f -> quoteIdentifier(f) + "=?")
+			.map(f -> format("%s = :%s", quoteIdentifier(f), f))
```
[the remark this answers is not in the document]
Comment (wuchong) [reply flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:102]: Using `VALUES(id)` might be more efficient than set values `:id` in statement. I would like to keep it. We can use the named variables in Oracle dialect.

**Comment by wuchong** [comment]
Thanks for the reviewing @leonardBang . I have addressed the review comments. I have to rebased the branch to resovle the compile problem.

**Comment by leonardBang** [comment]
Thanks @wuchong for the update, Postgres dialect tests failed because of it does use "`" as quote identifier,  please fix it and others looks good to me.

**Comment by leonardBang** [comment]
LGTM, +1 to merge

**Pull request closed without merging**